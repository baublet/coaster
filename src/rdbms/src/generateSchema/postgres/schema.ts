export const schema = [
  `DROP SCHEMA IF EXISTS public CASCADE;`,
  `CREATE SCHEMA public;`,
  `CREATE TABLE public.users (
      id serial PRIMARY KEY,
      username text UNIQUE NOT NULL,
      password text NOT NULL,
      email text UNIQUE NOT NULL,
      created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
      updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
  );`,
  `COMMENT ON TABLE public.users IS 'The users table';`,
  `DROP SCHEMA IF EXISTS accounts CASCADE;`,
  `CREATE SCHEMA accounts;`,
  `
    CREATE TABLE accounts.git_hub (
      id serial PRIMARY KEY,
      created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
      updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
      user_id integer NOT NULL REFERENCES public.users(id),
      github_id text UNIQUE NOT NULL
  );`,
  `COMMENT ON TABLE accounts.git_hub IS 'Connects GitHub accounts to user records';`,
];
