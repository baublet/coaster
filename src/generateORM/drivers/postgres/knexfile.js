module.exports = {
  useNullAsDefault: true,
  client: "pg",
  connection: {
    host: "127.0.0.1",
    port: 54311,
    user: "generate_orm_driver",
    password: "generate_orm_driver",
    database: "generate_orm_driver",
  },
};
