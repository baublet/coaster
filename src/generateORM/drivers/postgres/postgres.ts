import { RawEnum, RawSchema, RawTable } from "..";
import { shouldExclude, dbTypeToTypeScriptType } from "../helpers";
import { DatabaseConnection } from "../../";

interface PostgresOptions {
  /**
   * If not provided, we gather objects from only the `public` schema
   */
  schemas?: string[];
  /**
   * Table patterns to exclude. Uses `wildcard-match` under the hood.
   */
  excludeTables?: string[];
}

export const pgSchemaFetcher = async (
  connection: DatabaseConnection,
  options: PostgresOptions = {
    excludeTables: [],
  }
) => {
  const schemas = options.schemas || ["public"];
  const schemaTables: Record<string, string[]> = {};
  const tableComments: Map<string, string> = new Map();

  const queryResults = await connection
    .select<
      { table_name: string; table_schema: string; obj_description: string }[]
    >(
      connection.raw(
        "* , obj_description(('\"' || table_schema || '\"' || '.'|| '\"' || table_name || '\"')::regclass)"
      )
    )
    .from("information_schema.tables");
  for (const schema of schemas) {
    schemaTables[schema] = [];
    const resultTables = queryResults.filter(
      (table) => table.table_schema === schema
    );
    for (const table of resultTables) {
      schemaTables[schema].push(table.table_name);
      tableComments.set(
        `${table.table_schema}.${table.table_name}`,
        table.obj_description
      );
    }
  }

  const rawSchema: RawSchema[] = [];

  for (const [schemaName, tableNames] of Object.entries(schemaTables)) {
    const schemaTables: RawTable[] = [];
    for (const tableName of tableNames) {
      if (shouldExclude(options.excludeTables, tableName)) {
        continue;
      }

      const tableData = await connection
        .select<
          {
            column_name: string;
            column_default?: string | null;
            is_nullable: "YES" | "NO";
            data_type: string;
            is_updatable: "YES" | "NO";
            udt_name: string;
            col_description: string | null;
          }[]
        >(
          connection.raw(
            `*, col_description('"${schemaName}"."${tableName}"'::regclass, ordinal_position)`
          )
        )
        .from("information_schema.columns")
        .where("table_schema", "=", schemaName)
        .andWhere("table_name", "=", tableName);

      const tablePrimaryKeys: {
        rows: {
          attname: string;
          format_type: string;
        }[];
      } = await connection.raw(`
      SELECT
        pg_attribute.attname, 
        format_type(pg_attribute.atttypid, pg_attribute.atttypmod)
      FROM pg_index, pg_class, pg_attribute, pg_namespace 
      WHERE 
        pg_class.relname = '${tableName}' AND
        indrelid = pg_class.oid AND 
        nspname = '${schemaName}' AND 
        pg_class.relnamespace = pg_namespace.oid AND 
        pg_attribute.attrelid = pg_class.oid AND 
        pg_attribute.attnum = any(pg_index.indkey)
      AND indisprimary`);
      const primaryKey = tablePrimaryKeys.rows[0];

      const rawTable: RawTable = {
        name: tableName,
        columns: [],
        primaryKeyColumn: primaryKey.attname,
        primaryKeyType: dbTypeToTypeScriptType(primaryKey.format_type),
        comment: tableComments.get(`${schemaName}.${tableName}`),
      };

      for (const tableColumn of tableData) {
        const tsType = dbTypeToTypeScriptType(tableColumn.data_type);
        const enumPath =
          tsType === "enum"
            ? `${schemaName}.${tableColumn.udt_name}`
            : undefined;
        rawTable.columns.push({
          name: tableColumn.column_name,
          comment: tableColumn.col_description,
          nullable: tableColumn.is_nullable === "YES",
          hasDefault:
            tableColumn.column_default !== null ||
            tableColumn.column_default !== undefined,
          defaultTo: tableColumn.column_default,
          columnType: tableColumn.data_type,
          foreignKeys: [],
          uniqueConstraints: [],
          type: tsType,
          enumPath,
        });
      }

      const foreignKeys = await connection.raw<{
        rows: {
          constraint_name: string;
          column_name: string;
          foreign_table_name: string;
          foreign_table_schema: string;
          foreign_column_name: string;
        }[];
      }>(`
      SELECT
          tc.table_schema, 
          tc.constraint_name, 
          tc.table_name, 
          kcu.column_name,
          ccu.table_schema AS foreign_table_schema,
          ccu.table_name AS foreign_table_name,
          ccu.column_name AS foreign_column_name 
      FROM 
          information_schema.table_constraints AS tc 
          JOIN information_schema.key_column_usage AS kcu
            ON tc.constraint_name = kcu.constraint_name
            AND tc.table_schema = kcu.table_schema
          JOIN information_schema.constraint_column_usage AS ccu
            ON ccu.constraint_name = tc.constraint_name
            AND ccu.table_schema = tc.table_schema
      WHERE tc.constraint_type = 'FOREIGN KEY'
          AND tc.table_name='${tableName}'
          AND tc.table_schema='${schemaName}';
          `);

      for (const rawConstraint of foreignKeys.rows) {
        const column = rawTable.columns.find(
          (c) => c.name === rawConstraint.column_name
        );
        if (!column) {
          // TODO: handle where a column has a constraint that points to non-existent column
          continue;
        }
        column.foreignKeys.push({
          column: rawConstraint.foreign_column_name,
          table: rawConstraint.foreign_table_name,
          schema: rawConstraint.foreign_table_schema,
        });
      }

      // TODO: load the unique constraints

      schemaTables.push(rawTable);
    }

    // Enums
    const enums = await connection.raw<{
      rows: {
        enum_name: string;
        enum_value: string;
      }[];
    }>(`SELECT t.typname AS enum_name,
    e.enumlabel AS enum_value
  FROM pg_type t 
    JOIN pg_enum e on t.oid = e.enumtypid  
    JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
  WHERE n.nspname = '${schemaName}'`);
    const enumsMap = new Map<string, string[]>();
    for (const { enum_name, enum_value } of enums.rows) {
      if (!enumsMap.has(enum_name)) {
        enumsMap.set(enum_name, []);
      }
      enumsMap.get(enum_name).push(enum_value);
    }

    const schemaEnums: RawEnum[] = [];
    for (const [enumName, enumValues] of enumsMap) {
      schemaEnums.push({
        name: enumName,
        values: enumValues,
      });
    }

    rawSchema.push({
      flavor: "pg",
      name: schemaName,
      tables: schemaTables,
      enums: schemaEnums,
    });
  }

  return rawSchema;
};
