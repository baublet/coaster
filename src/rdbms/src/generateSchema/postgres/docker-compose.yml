version: "3.1"

services:
  db:
    image: postgres:9.6-alpine
    restart: always
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
    ports:
      - "5438:5432"
