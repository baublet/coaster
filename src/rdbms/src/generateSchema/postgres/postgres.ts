import knex from "knex";
import match from "wildcard-match";

import { CoasterError, createCoasterError } from "@baublet/coaster-utils";

import {
  GenerateCoasterRdbmsConnectionOptions,
  RdbmsSchema,
  RdbmsTable,
  RdbmsColumn,
} from "../../types";
import { transformName } from "../transformName";

type QueryFunction = <T>(query: string, params?: any[]) => Promise<T>;

const SCHEMA_TO_IGNORE = ["information_schema", "pg_*"];
const isIgnored = match(SCHEMA_TO_IGNORE);

export async function postgres(
  options: GenerateCoasterRdbmsConnectionOptions
): Promise<RdbmsSchema | CoasterError> {
  const connection = knex(options.config);
  const query: QueryFunction = async (query: string, params: any[] = []) => {
    const results = await connection.raw(query, params);
    return results.rows;
  };

  const schemas: string[] = [];

  const { indexes } = await getSystemIndexes({
    query,
    options,
  });

  if (options.schemas) {
    schemas.push(...options.schemas);
  } else {
    const result = await query<{ schema_name: string }[]>(
      "SELECT * FROM information_schema.schemata"
    );

    schemas.push(
      ...result
        .filter((row) => !isIgnored(row.schema_name))
        .map((row) => row.schema_name)
    );
  }

  if (schemas.length === 0) {
    return createCoasterError({
      code: "generateSchema-postgres-noSchemas",
      message: "No schemas found in provided postgres connection",
      details: { options },
    });
  }

  const allTables: RdbmsTable[] = [];
  const schemaTablePromises: Promise<void>[] = [];
  for (const schema of schemas) {
    schemaTablePromises.push(
      attachSchemaTables({ query, schema, allTables, options, indexes })
    );
  }
  await Promise.all(schemaTablePromises);

  const schema: RdbmsSchema = {
    tables: allTables,
  };

  return schema;
}

async function attachSchemaTables({
  query,
  schema,
  allTables,
  options,
  indexes,
}: {
  schema: string;
  query: QueryFunction;
  allTables: RdbmsTable[];
  options: GenerateCoasterRdbmsConnectionOptions;
  indexes: Map<string, boolean>;
}): Promise<void> {
  const results = await query<
    { table_name: string; table_schema: string; comment?: string }[]
  >(
    `
      SELECT
        t.table_name,
        t.table_schema,
        pg_catalog.obj_description(pgc.oid, 'pg_class') as comment
      FROM information_schema.tables t
      INNER JOIN pg_catalog.pg_class pgc
      ON t.table_name = pgc.relname 
      WHERE t.table_type = 'BASE TABLE'
        AND t.table_schema = ?;
    `,
    [schema]
  );

  const tablePromises: Promise<void>[] = [];
  for (const {
    table_name: tableName,
    table_schema: schemaName,
    comment,
  } of results) {
    const tableNiceName = transformName({
      name: schemaName === "public" ? tableName : `${schemaName}_${tableName}`,
      transformer: options.namingConventions?.tableName,
      defaultTransformer: "pascalCase",
    });

    tablePromises.push(
      new Promise((resolve) => {
        const table: RdbmsTable = {
          name: tableNiceName,
          nameInDb: `"${schemaName}"."${tableName}"`,
          columns: [],
          comment,
        };
        allTables.push(table);

        tablePromises.push(
          attachTableColumns({
            query,
            table,
            options,
            schemaName,
            tableName,
            indexes,
          })
        );

        resolve();
      })
    );
  }

  await Promise.all(tablePromises);
}

async function attachTableColumns({
  query,
  table,
  options,
  schemaName,
  tableName,
  indexes,
}: {
  table: RdbmsTable;
  query: QueryFunction;
  options: GenerateCoasterRdbmsConnectionOptions;
  schemaName: string;
  tableName: string;
  indexes: Map<string, boolean>;
}): Promise<void> {
  const columns = await query<
    {
      defaultValue: string | null;
      columnName: string;
      isNullable: "YES" | "NO";
      dataType: string;
      columnDefault: string | null;
      isUpdatable: "YES" | "NO";
    }[]
  >(
    `
      SELECT column_default AS "defaultValue",
        column_name AS "columnName",
        data_type AS "dataType",
        is_nullable AS "isNullable",
        is_updatable AS "isUpdatable"
      FROM information_schema.columns
      WHERE table_schema = ?
        AND table_name = ?
      ORDER BY ordinal_position;
  `,
    [schemaName, tableName]
  );

  for (const {
    defaultValue,
    columnName,
    dataType,
    isNullable,
    isUpdatable,
  } of columns) {
    const columnNiceName = transformName({
      name: columnName,
      transformer: options.namingConventions?.columnName,
      defaultTransformer: "camelCase",
    });
    const nullable = isNullable === "YES";

    const column: RdbmsColumn = {
      name: columnNiceName,
      nameInDb: `"${columnName}"`,
      type: dataType,
      nullable,
      updatable: isUpdatable === "YES",
      indexed: Boolean(indexes.get(`${schemaName}.${tableName}.${columnName}`)),
      hasDefaultValue: defaultValue !== null,
      valueType: getTypeScriptTypeFromPostgresColumnType({
        dataType,
        nullable,
      }),
    };

    table.columns.push(column);
  }
}

async function getSystemIndexes({
  query,
}: {
  query: QueryFunction;
  options: GenerateCoasterRdbmsConnectionOptions;
}): Promise<{
  indexes: Map<string, boolean>;
  data: {
    schemaName: string;
    tableName: string;
    columnNames: string[];
    isPrimary: boolean;
  }[];
}> {
  const results = await query<
    {
      schemaName: string;
      tableName: string;
      columnNames: string[];
      isUnique: boolean;
      isPrimary: boolean;
    }[]
  >(
    `
    SELECT
      U.usename                AS "userName",
      ns.nspname               AS "schemaName",
      idx.indrelid :: REGCLASS AS "tableName",
      i.relname                AS "indexName",
      idx.indisunique          AS "isUnique",
      idx.indisprimary         AS "isPrimary",
      am.amname                AS "indexType",
      idx.indkey,
          ARRAY(
              SELECT pg_get_indexdef(idx.indexrelid, k + 1, TRUE)
              FROM
                generate_subscripts(idx.indkey, 1) AS k
              ORDER BY k
          ) AS "columnNames",
      (idx.indexprs IS NOT NULL) OR (idx.indkey::int[] @> array[0]) AS "isFunctional",
      idx.indpred IS NOT NULL AS "isPartial"
    FROM pg_index AS idx
      JOIN pg_class AS i
        ON i.oid = idx.indexrelid
      JOIN pg_am AS am
        ON i.relam = am.oid
      JOIN pg_namespace AS NS ON i.relnamespace = NS.OID
      JOIN pg_user AS U ON i.relowner = U.usesysid
    WHERE NOT nspname LIKE 'pg%'; -- Excluding system tables
    `
  );

  const data = results.map((result) => ({
    schemaName: result.schemaName,
    tableName: result.tableName,
    isPrimary: result.isPrimary,
    columnNames: result.columnNames,
  }));

  const indexes = new Map<string, boolean>();

  for (const { schemaName, tableName, columnNames } of data) {
    for (const columnName of columnNames) {
      indexes.set(`${schemaName}.${tableName}.${columnName}`, true);
    }
  }

  return { indexes, data };
}

function getTypeScriptTypeFromPostgresColumnType({
  dataType,
  nullable,
}: {
  dataType: string;
  nullable: boolean;
}): string {
  const pieces: string[] = nullable ? ["null", "undefined"] : [];

  if (
    dataType.includes("int") ||
    dataType.includes("real") ||
    dataType.includes("numeric") ||
    dataType.includes("decimal") ||
    dataType.includes("double") ||
    dataType.includes("money") ||
    dataType.includes("serial")
  ) {
    pieces.push("number");
  } else if (
    dataType.includes("char") ||
    dataType.includes("text") ||
    dataType.includes("uuid")
  ) {
    pieces.push("string");
  } else if (dataType.includes("bool")) {
    pieces.push("boolean");
  } else if (
    dataType.includes("date") ||
    dataType.includes("time") ||
    dataType.includes("interval")
  ) {
    pieces.push("Date");
  } else if (dataType.includes("json")) {
    pieces.push("any");
  } else {
    pieces.push("unknown");
  }

  return pieces.join(" | ");
}
