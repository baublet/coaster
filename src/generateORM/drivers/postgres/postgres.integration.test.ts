import knex from "knex";

import { pgSchemaFetcher } from "./postgres";

const connection = knex(require("./knexfile.js"));

afterAll(async () => {
  await connection.destroy();
});

it("returns a schema in the right shape", async () => {
  await expect(pgSchemaFetcher(connection)).resolves.toMatchSnapshot();
});

it("excludes things when asked", async () => {
  await expect(
    pgSchemaFetcher(connection, {
      excludeTables: ["knex_*"],
    })
  ).resolves.toMatchSnapshot();
});
