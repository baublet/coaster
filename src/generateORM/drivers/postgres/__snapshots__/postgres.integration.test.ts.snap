// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excludes things when asked 1`] = `
Array [
  Object {
    "enums": Array [
      Object {
        "name": "source",
        "values": Array [
          "github",
          "bitbucket",
        ],
      },
    ],
    "flavor": "pg",
    "name": "public",
    "tables": Array [
      Object {
        "columns": Array [
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "id",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "created_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "updated_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": "Users can have multiple accounts",
        "name": "users",
        "primaryKeyColumn": "id",
        "primaryKeyType": "string",
      },
      Object {
        "columns": Array [
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "id",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "created_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "updated_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "USER-DEFINED",
            "comment": null,
            "enumPath": "public.source",
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "source",
            "nullable": false,
            "type": "enum",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "uniqueIdentifier",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "text",
            "comment": "Users can have multiple accounts",
            "enumPath": undefined,
            "foreignKeys": Array [
              Object {
                "column": "id",
                "schema": "public",
                "table": "users",
              },
            ],
            "hasDefault": false,
            "name": "userId",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "jsonb",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "rawUserData",
            "nullable": true,
            "type": "AnyJson",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": "UserAccount table comment",
        "name": "userAccounts",
        "primaryKeyColumn": "id",
        "primaryKeyType": "string",
      },
    ],
  },
]
`;

exports[`returns a schema in the right shape 1`] = `
Array [
  Object {
    "enums": Array [
      Object {
        "name": "source",
        "values": Array [
          "github",
          "bitbucket",
        ],
      },
    ],
    "flavor": "pg",
    "name": "public",
    "tables": Array [
      Object {
        "columns": Array [
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "id",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "created_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "updated_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": "Users can have multiple accounts",
        "name": "users",
        "primaryKeyColumn": "id",
        "primaryKeyType": "string",
      },
      Object {
        "columns": Array [
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "id",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "created_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "updated_at",
            "nullable": false,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "USER-DEFINED",
            "comment": null,
            "enumPath": "public.source",
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "source",
            "nullable": false,
            "type": "enum",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "text",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "uniqueIdentifier",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "text",
            "comment": "Users can have multiple accounts",
            "enumPath": undefined,
            "foreignKeys": Array [
              Object {
                "column": "id",
                "schema": "public",
                "table": "users",
              },
            ],
            "hasDefault": false,
            "name": "userId",
            "nullable": false,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "jsonb",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "rawUserData",
            "nullable": true,
            "type": "AnyJson",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": "UserAccount table comment",
        "name": "userAccounts",
        "primaryKeyColumn": "id",
        "primaryKeyType": "string",
      },
      Object {
        "columns": Array [
          Object {
            "columnType": "integer",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "id",
            "nullable": false,
            "type": "number",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "character varying",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "name",
            "nullable": true,
            "type": "string",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "integer",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "batch",
            "nullable": true,
            "type": "number",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "timestamp with time zone",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "migration_time",
            "nullable": true,
            "type": "Date",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": null,
        "name": "knex_migrations",
        "primaryKeyColumn": "id",
        "primaryKeyType": "number",
      },
      Object {
        "columns": Array [
          Object {
            "columnType": "integer",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": true,
            "name": "index",
            "nullable": false,
            "type": "number",
            "uniqueConstraints": Array [],
          },
          Object {
            "columnType": "integer",
            "comment": null,
            "enumPath": undefined,
            "foreignKeys": Array [],
            "hasDefault": false,
            "name": "is_locked",
            "nullable": true,
            "type": "number",
            "uniqueConstraints": Array [],
          },
        ],
        "comment": null,
        "name": "knex_migrations_lock",
        "primaryKeyColumn": "index",
        "primaryKeyType": "number",
      },
    ],
  },
]
`;
